### ISSUE TITLE
espanso 2.0.4-alpha, Arch Linux, Wayland: read wayland display from env variable instead of hard-coding it

### ISSUE THREAD

**Issue by synaptiko** [federico-terzi/espanso#846]
Hello, I've followed the steps in https://espanso.org/docs/next/install/linux/#install-on-wayland, including setting the capabilities and creating a config file with a layout but when espanso starts, it's stops almost immediately.

Here's the log:
```
19:49:27 [daemon(116705)] [INFO] reading configs from: "/home/eggze/.config/espanso"
19:49:27 [daemon(116705)] [INFO] reading packages from: "/home/eggze/.config/espanso/match/packages"
19:49:27 [daemon(116705)] [INFO] using runtime dir: "/home/eggze/.cache/espanso"
19:49:27 [daemon(116705)] [WARN] unable to determine the currently active layout, you might need to explicitly specify the layout in the config for espanso to work correctly.
19:49:27 [daemon(116705)] [WARN] keyboard layout watcher couldn't determine active layout.
19:49:27 [daemon(116705)] [INFO] watching for changes in path: "/home/eggze/.config/espanso"
19:49:27 [daemon(116705)] [INFO] espanso version: 2.0.3-alpha
19:49:27 [daemon(116705)] [INFO] spawning the worker process...
19:49:27 [daemon(116705)] [INFO] binded to IPC unix socket: /home/eggze/.cache/espanso/espansodaemonv2.sock
19:49:27 [worker(116711)] [INFO] reading configs from: "/home/eggze/.config/espanso"
19:49:27 [worker(116711)] [INFO] reading packages from: "/home/eggze/.config/espanso/match/packages"
19:49:27 [worker(116711)] [INFO] using runtime dir: "/home/eggze/.cache/espanso"
19:49:27 [worker(116711)] [INFO] binded to IPC unix socket: /home/eggze/.cache/espanso/espansoworkerv2.sock
19:49:27 [worker(116711)] [INFO] using WaylandAppInfoProvider
19:49:27 [worker(116711)] [INFO] detection module will use this keyboard layout: [R=, M=, L=it, V=, O=]
19:49:27 [worker(116711)] [INFO] monitoring the status of the daemon process
19:49:27 [worker(116711)] [INFO] using EVDEVSource
19:49:28 [worker(116711)] [INFO] Querying modifier status...
19:49:28 [worker(116711)] [INFO] inject module will use this keyboard layout: [R=, M=, L=it, V=, O=]
19:49:28 [worker(116711)] [INFO] using EVDEVInjector
19:49:28 [worker(116711)] [ERROR] thread 'engine thread' panicked at 'failed to initialize injector module: could not open uinput device': espanso/src/cli/worker/engine/mod.rs:174
19:49:28 [worker(116711)] [ERROR] Unable to block the LinuxEventLoop: receiving on an empty and disconnected channel
19:49:28 [worker(116711)] [ERROR] thread 'main' panicked at 'unable to run main eventloop: receiving on an empty and disconnected channel': espanso/src/cli/worker/mod.rs:155
19:49:28 [daemon(116705)] [WARN] unable to determine the currently active layout, you might need to explicitly specify the layout in the config for espanso to work correctly.
19:49:28 [daemon(116705)] [ERROR] received unexpected exit code from worker 101, exiting
```

I've updated `~/.config/espanso/config/default.yml` with:
```
keyboard_layout:
  layout: "it"
```
but the log above still says it cannot recognize the layout.

Also `sudo setcap "cap_dac_override+p" /usr/local/bin/espanso` finished with success but there is `could not open uinput device`.

I think that the path to `espanso` is correct, just checked with:
```
❯ which espanso
/usr/local/bin/espanso
```

I didn't use `espanso` before.

**Comment by synaptiko** [comment]
I'm using basic Arch Linux setup.

The result is:
```
❯ ls -la /dev/uinput
crw-rw----+ 1 root root 10, 223 04.11.2021 06:53 /dev/uinput
```

I also tried:
```
❯ sudo modprobe uinput
modprobe: FATAL: Module uinput not found in directory /lib/modules/5.14.15-arch1-1
```
and immediately realized that I recently upgraded kernel. I'll try it once again after reboot. Sorry, it seems the issue may be on my side :man_facepalming:

**Comment by synaptiko** [comment]
The error from the log is gone after reboot but espanso still doesn't work. I can see it starts, some window blinks and then it stops. This is `journalctl`'s output:
```
Nov 04 21:27:44 jprokop-yoga systemd[795]: Started espanso.
Nov 04 21:27:44 jprokop-yoga espanso[1735]: 21:27:44 [daemon(1735)] [INFO] reading configs from: "/home/eggze/.config/espanso"
Nov 04 21:27:44 jprokop-yoga espanso[1735]: 21:27:44 [daemon(1735)] [INFO] reading packages from: "/home/eggze/.config/espanso/match/packages"
Nov 04 21:27:44 jprokop-yoga espanso[1735]: 21:27:44 [daemon(1735)] [INFO] using runtime dir: "/home/eggze/.cache/espanso"
Nov 04 21:27:44 jprokop-yoga espanso[1735]: 21:27:44 [daemon(1735)] [WARN] unable to determine the currently active layout, you might need to explicit>
Nov 04 21:27:44 jprokop-yoga espanso[1735]: 21:27:44 [daemon(1735)] [WARN] keyboard layout watcher couldn't determine active layout.
Nov 04 21:27:44 jprokop-yoga espanso[1735]: 21:27:44 [daemon(1735)] [INFO] watching for changes in path: "/home/eggze/.config/espanso"
Nov 04 21:27:44 jprokop-yoga espanso[1735]: 21:27:44 [daemon(1735)] [INFO] espanso version: 2.0.3-alpha
Nov 04 21:27:44 jprokop-yoga espanso[1735]: 21:27:44 [daemon(1735)] [INFO] spawning the worker process...
Nov 04 21:27:44 jprokop-yoga espanso[1735]: 21:27:44 [daemon(1735)] [INFO] binded to IPC unix socket: /home/eggze/.cache/espanso/espansodaemonv2.sock
Nov 04 21:27:44 jprokop-yoga espanso[1741]: 21:27:44 [worker(1741)] [INFO] reading configs from: "/home/eggze/.config/espanso"
Nov 04 21:27:44 jprokop-yoga espanso[1741]: 21:27:44 [worker(1741)] [INFO] reading packages from: "/home/eggze/.config/espanso/match/packages"
Nov 04 21:27:44 jprokop-yoga espanso[1741]: 21:27:44 [worker(1741)] [INFO] using runtime dir: "/home/eggze/.cache/espanso"
Nov 04 21:27:44 jprokop-yoga espanso[1741]: 21:27:44 [worker(1741)] [INFO] binded to IPC unix socket: /home/eggze/.cache/espanso/espansoworkerv2.sock
Nov 04 21:27:44 jprokop-yoga espanso[1741]: 21:27:44 [worker(1741)] [INFO] using WaylandAppInfoProvider
Nov 04 21:27:44 jprokop-yoga espanso[1741]: 21:27:44 [worker(1741)] [INFO] detection module will use this keyboard layout: [R=, M=, L=it, V=, O=]
Nov 04 21:27:44 jprokop-yoga espanso[1741]: 21:27:44 [worker(1741)] [INFO] monitoring the status of the daemon process
Nov 04 21:27:44 jprokop-yoga espanso[1741]: 21:27:44 [worker(1741)] [INFO] using EVDEVSource
Nov 04 21:27:44 jprokop-yoga systemd[795]: Stopping espanso...
Nov 04 21:27:45 jprokop-yoga systemd[795]: Stopped espanso.
```

**Comment by federico-terzi** [comment]
@synaptiko Interesting, thank you! Could you please try running the following command, then stop it after a while (5/6 seconds) and post the output here?

```
espanso -vv worker
```

Thanks!

**Comment by synaptiko** [comment]
Here it is:
```
❯ espanso -vv worker
07:21:50 [worker(5339)] [DEBUG] (1) espanso_path: detected config directory in $HOME/.config/espanso
07:21:50 [worker(5339)] [DEBUG] (1) espanso_path: detected default runtime dir: "/home/eggze/.cache/espanso"
07:21:50 [worker(5339)] [DEBUG] (1) espanso_path: detected default packages dir: "/home/eggze/.config/espanso/match/packages"
07:21:50 [worker(5339)] [INFO] reading configs from: "/home/eggze/.config/espanso"
07:21:50 [worker(5339)] [INFO] reading packages from: "/home/eggze/.config/espanso/match/packages"
07:21:50 [worker(5339)] [INFO] using runtime dir: "/home/eggze/.cache/espanso"
07:21:50 [worker(5339)] [DEBUG] (1) espanso_config::config::store: loaded default config at path: "/home/eggze/.config/espanso/config/default.yml"
07:21:50 [worker(5339)] [DEBUG] (1) espanso::patch::config_store: skipping 'alacritty_terminal_x11' patch
07:21:50 [worker(5339)] [DEBUG] (1) espanso::patch::config_store: skipping 'emacs_x11' patch
07:21:50 [worker(5339)] [DEBUG] (1) espanso::patch::config_store: skipping 'generic_terminal_x11' patch
07:21:50 [worker(5339)] [DEBUG] (1) espanso::patch::config_store: skipping 'kitty_terminal_x11' patch
07:21:50 [worker(5339)] [DEBUG] (1) espanso::patch::config_store: skipping 'konsole_terminal_x11' patch
07:21:50 [worker(5339)] [DEBUG] (1) espanso::patch::config_store: skipping 'libreoffice_writer_x11' patch
07:21:50 [worker(5339)] [DEBUG] (1) espanso::patch::config_store: skipping 'simple_terminal_x11' patch
07:21:50 [worker(5339)] [DEBUG] (1) espanso::patch::config_store: skipping 'simple_terminal_2_x11' patch
07:21:50 [worker(5339)] [DEBUG] (1) espanso::patch::config_store: skipping 'terminator_terminal_x11' patch
07:21:50 [worker(5339)] [DEBUG] (1) espanso::patch::config_store: skipping 'termite_terminal_x11' patch
07:21:50 [worker(5339)] [DEBUG] (1) espanso::patch::config_store: skipping 'thunderbird_x11' patch
07:21:50 [worker(5339)] [DEBUG] (1) espanso::patch::config_store: skipping 'tilix_terminal_x11' patch
07:21:50 [worker(5339)] [DEBUG] (1) espanso::patch::config_store: skipping 'urxvt_terminal_x11' patch
07:21:50 [worker(5339)] [DEBUG] (1) espanso::patch::config_store: skipping 'xterm_terminal_x11' patch
07:21:50 [worker(5339)] [DEBUG] (1) espanso::patch::config_store: skipping 'yakuake_terminal_x11' patch
07:21:50 [worker(5339)] [DEBUG] (1) espanso::cli::worker: starting with start-reason = None
07:21:50 [worker(5339)] [DEBUG] (1) espanso::icon: skipping extraction for '"/home/eggze/.cache/espanso/iconv2.png"', as it's already present
07:21:50 [worker(5339)] [DEBUG] (1) espanso::icon: skipping extraction for '"/home/eggze/.cache/espanso/search.png"', as it's already present
07:21:50 [worker(5339)] [DEBUG] (1) espanso::icon: skipping extraction for '"/home/eggze/.cache/espanso/icon_no_background.png"', as it's already present
07:21:50 [worker(5339)] [INFO] using WaylandAppInfoProvider
07:21:50 [worker(5339)] [INFO] binded to IPC unix socket: /home/eggze/.cache/espanso/espansoworkerv2.sock
07:21:50 [worker(5339)] [DEBUG] (2) espanso::cli::worker::engine: using linux capabilities to grant permissions needed by EVDEV backend
07:21:50 [worker(5339)] [DEBUG] (2) espanso::cli::worker::engine: successfully granted permissions using capabilities
07:21:50 [worker(5339)] [INFO] detection module will use this keyboard layout: [R=, M=, L=it, V=, O=]
07:21:50 [worker(5339)] [INFO] using EVDEVSource
07:21:50 [worker(5339)] [INFO] Querying modifier status...
07:21:50 [worker(5339)] [DEBUG] (4) espanso_detect::evdev: Updating device modifier state: ModifiersState { ctrl: false, alt: false, shift: false, caps_lock: false, meta: false, num_lock: false }
07:21:50 [worker(5339)] [INFO] inject module will use this keyboard layout: [R=, M=, L=it, V=, O=]
07:21:50 [worker(5339)] [INFO] using EVDEVInjector
07:21:51 [worker(5339)] [INFO] using WaylandFallbackClipboard
07:21:51 [worker(5339)] [ERROR] failed to connect to Wayland display
07:21:51 [worker(5339)] [ERROR] thread 'engine thread' panicked at 'failed to initialize clipboard module: can't connect to Wayland display': espanso/src/cli/worker/engine/mod.rs:176
07:21:51 [worker(5339)] [ERROR] Unable to block the LinuxEventLoop: receiving on an empty and disconnected channel
07:21:51 [worker(5339)] [ERROR] thread 'main' panicked at 'unable to run main eventloop: receiving on an empty and disconnected channel': espanso/src/cli/worker/mod.rs:155
```

I also ran:
```
❯ env | grep WAYLAND_DISPLAY
WAYLAND_DISPLAY=wayland-1
```
I think my env is correct. To clarify, I'm using Sway as a user service under systemd and I believe I have correct exports as well, because other apps which are started over dbus or systemd are working fine.

**Comment by synaptiko** [comment]
Can it be that `wayland-0` is hard-coded here? https://github.com/federico-terzi/espanso/blob/aa9465490b141d8816f49d10328aa3836cfe3902/espanso-clipboard/src/wayland/fallback/mod.rs#L52

**Comment by federico-terzi** [comment]
@synaptiko Interesting! So the `uinput` error was solved after the reboot

> Update: by running ln -s /run/user/1001/wayland-1 /run/user/1001/wayland-0 I made espanso to start.

That's nice! Instead of hardcoding `wayland-0` we should read it from the env variable as you suggested, I'll update the ticket :) Thanks

**Comment by federico-terzi** [comment]
Hey,

I'm publishing a fix in the upcoming v2.0.5-alpha release, thank you for the help!

Cheers :)
